### ISSUE TITLE
tlslite with tornado

### ISSUE THREAD

**Issue by hamdell** [tomato42/tlslite-ng#381]
Hi,I want to wirte httpserver，can tlslite with torado？

**Comment by tomato42** [comment]
I'm not familiar with torado, and don't see anything in the code referencing it, so I'd say that the answer as of now is "no".

That being said, it shouldn't be too hard, there already are integrations with few other modules, see the `tlslite/integration` directory, e.g.: https://github.com/tomato42/tlslite-ng/blob/master/tlslite/integration/httptlsconnection.py

**Comment by hamdell** [comment]
@tomato42 ,Thank you for your reply. That's not what I want. I want a highly concurrent and asynchronous HTTP server to get SSL clienthello sent by the browser and respond to the browser in the form of HTTP-JSON. Can you write it?

**Comment by tomato42** [comment]
> I want a highly concurrent and asynchronous [...] server

while this library does support asynchronous operation, it's definitely not fast

> HTTP server to [...] respond to the browser in the form of HTTP-JSON

once you establish a TLS connection you can exchange anything you want on top of it...

> Can you write it?

no, that's not why I'm working on this library

**Comment by hamdell** [comment]
@tomato42 ,thanks，a highly concurrent httpserver to get SSL clienthello sent by the browser is very important to me,the sooner the better,i hope it!

**Comment by hamdell** [comment]
@tomato42 ,I'm very sorry. I misspelled it. It's tornado, a python's high concurrency asynchronous web framework. I saw / scripts / tls.py. So, can tornado replace basehttpserver?

**Comment by tomato42** [comment]
> So, can tornado replace basehttpserver?

I'd be surprised if it couldn't. If somebody proposes necessary changes to integrate tlslite-ng with tornado, I will review them, but I won't work on them myself.

**Comment by hamdell** [comment]
@tomato42 ,Yes,thank you for your reply. Do you have time to call tornado to write an httpserver script similar to tls.py as soon as possible? It's very important to me, thank you very much!

**Comment by HMaker** [comment]
All good async frameworks integrates well with standard python [SSL APIs][1], you need to reimplement the APIs tornado uses using TLSLite as base. TLSLite supports non-blocking sockets which will be used by tornado, its async functions are generators which yields 0 when it wants a readable socket and 1 for writable socket, most async event loops has low level APIs which allows you to set callbacks for when sockets become readable and writable, use it to resume TLSLite generators.

[1]: https://docs.python.org/3.11/library/ssl.html
